### PR TITLE
Harmpack cuff fix

### PR DESCRIPTION
**Pull request by insoPL**
## About the PR

Fixes 2 bugs with cuff interactions with harmpack.

Both bugs are from this issue here:
https://github.com/space-wizards/space-station-14/issues/39253

## Requirements

- [x] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [x] I have added media to this PR or it does not require an ingame showcase.

## Breaking changes

        public bool TryAddNewCuffs(EntityUid target, EntityUid user, EntityUid handcuff, CuffableComponent? component = null, HandcuffComponent? cuff = null)
        
        changed into
        
         public bool TryAddNewCuffs(Entity<CuffableComponent?> target, EntityUid user, EntityUid handcuff)       
         cuff was not used so it was yeeted and target is Entity<Comp>
         
**Changelog**

### CODE REVIEW

File: Content.Shared/Cuffs/SharedCuffableSystem.cs [hunk 1]
```diff
@@ -122,13 +122,14 @@ private void OnUncuffAttempt(ref UncuffAttemptEvent args)
                 // We temporarily allow interactions so the cuffable system does not block itself.
                 // It's assumed that this will always be false.
                 // Otherwise they would not be trying to uncuff themselves.
```
Comment (LordCarve) [inline Content.Shared/Cuffs/SharedCuffableSystem.cs]: Latter two lines of this comment need an update.
It's obviously no longer assumed to be false.

Just removing them them seems good enough, but it's up to you.

**Comment by LordCarve** [comment]
After testing locally this:
- addresses some problems with harmpack + handcuffs interactions,
- doesn't address all problems that harmpack + handcuffs interactions have,
- doesn't affect regular handcuffs (no-harmpack) interactions
- doesn't introduce any new bad harmpack + handcuffs interactions
- makes the "still bugged" interactions consistent (i.e. taking actions such as trying to unrestrain and get interrupted won't "deepen" the bug, but leave it as bugged as it was)

So overall this PR is a net positive, even if it doesn't address all of the issues.

Details to follow:

--------------

How this PR affects problems described in linked issue:
- Missing cuffed icon when double-cuffed with harmpack = ✅ fixed
- Missing cuffed icon when cuffed + harmpack is removed = ✅ fixed
- "fake blocked by handcuffs" after removing harmpack and cuffed = ❔ unable to replicate before or after
- Missing cuffed icon when single-cuffed with harmpack ❌ icon still doesn't show
- Uncuffed hands usage blocked if unrestraining interrupted when single-cuffed with harmpack = ✅ fixed

I also found many more other buggy interactions with handcuffs + harmpacks.
Some of them are non-harmpack but just many-handed mob issues (>2), but harmpacks letting us add and remove hands at will exacerbate the issue further.

If you don't want to fix them in this PR let me know I will create a new issue for them:
- Which hands end up restrained if more than two are available is random (controlled if self-restraining) and produces different effects:
- Can handcuff one harm-hand with one-regular hand and upon dropping harmpack having one handcuffed hand and the other free but unusable ❌
- Cannot use harm-hands if harmpack put on after regular hands restrained ❌ but if another set is put on and removed it works
- Cannot use any hands if their regular hands get restrained (gets cuffed icon), but can use any free hands if at least one regular hand left unrestrained ❌

Seems like this in general needs a more thorough fix where restrainment status is divorced from hands usability and both dynamically depend on number of restrained/free hands. In short: restrainment should be per hand, and statuses derived from restrainment (icon, inability to switch hands) should depend on hands' statuses instead of being their own value.

**Comment by insoPL** [comment]
> After testing locally this:
> - addresses some problems with harmpack + handcuffs interactions,
> - doesn't address all problems that harmpack + handcuffs interactions have,
> - doesn't affect regular handcuffs (no-harmpack) interactions
> - doesn't introduce any new bad harmpack + handcuffs interactions
> - makes the "still bugged" interactions consistent (i.e. taking actions such as trying to unrestrain and get interrupted won't "deepen" the bug, but leave it as bugged as it was)
> 
> So overall this PR is a net positive, even if it doesn't address all of the issues.
> 
> Details to follow:
> 
> --------------
> 
> How this PR affects problems described in linked issue:
> - Missing cuffed icon when double-cuffed with harmpack = ✅ fixed
> - Missing cuffed icon when cuffed + harmpack is removed = ✅ fixed
> - "fake blocked by handcuffs" after removing harmpack and cuffed = ❔ unable to replicate before or after
> - Missing cuffed icon when single-cuffed with harmpack ❌ icon still doesn't show
> - Uncuffed hands usage blocked if unrestraining interrupted when single-cuffed with harmpack = ✅ fixed
> 
> I also found many more other buggy interactions with handcuffs + harmpacks.
> Some of them are non-harmpack but just many-handed mob issues (>2), but harmpacks letting us add and remove hands at will exacerbate the issue further.
> 
> If you don't want to fix them in this PR let me know I will create a new issue for them:
> - Which hands end up restrained if more than two are available is random (controlled if self-restraining) and produces different effects:
> - Can handcuff one harm-hand with one-regular hand and upon dropping harmpack having one handcuffed hand and the other free but unusable ❌
> - Cannot use harm-hands if harmpack put on after regular hands restrained ❌ but if another set is put on and removed it works
> - Cannot use any hands if their regular hands get restrained (gets cuffed icon), but can use any free hands if at least one regular hand left unrestrained ❌
> 
> Seems like this in general needs a more thorough fix where restrainment status is divorced from hands usability and both dynamically depend on number of restrained/free hands. In short: restrainment should be per hand, and statuses derived from restrainment (icon, inability to switch hands) should depend on hands' statuses instead of being their own value.

Thanks for testing.

"Missing cuffed icon when single-cuffed with harmpack ❌ icon still doesn't show" I am not sure if that is bug or intended behaviour.

I think Cuffing should work always in order:  normal arms then, additional arms. Another problem is future proofing for 1 and 3 arm entities(surgery soon). I will bother maitainers on discord about it when i have some spare time.

**Comment by insoPL** [comment]
> After testing locally this:
> 
> * addresses some problems with harmpack + handcuffs interactions,
> * doesn't address all problems that harmpack + handcuffs interactions have,
> * doesn't affect regular handcuffs (no-harmpack) interactions
> * doesn't introduce any new bad harmpack + handcuffs interactions
> * makes the "still bugged" interactions consistent (i.e. taking actions such as trying to unrestrain and get interrupted won't "deepen" the bug, but leave it as bugged as it was)
> 
> So overall this PR is a net positive, even if it doesn't address all of the issues.
> 
> Details to follow:
> 
> How this PR affects problems described in linked issue:
> 
> * Missing cuffed icon when double-cuffed with harmpack = ✅ fixed
> * Missing cuffed icon when cuffed + harmpack is removed = ✅ fixed
> * "fake blocked by handcuffs" after removing harmpack and cuffed = ❔ unable to replicate before or after
> * Missing cuffed icon when single-cuffed with harmpack ❌ icon still doesn't show
> * Uncuffed hands usage blocked if unrestraining interrupted when single-cuffed with harmpack = ✅ fixed
> 
> I also found many more other buggy interactions with handcuffs + harmpacks. Some of them are non-harmpack but just many-handed mob issues (>2), but harmpacks letting us add and remove hands at will exacerbate the issue further.
> 
> If you don't want to fix them in this PR let me know I will create a new issue for them:
> 
> * Which hands end up restrained if more than two are available is random (controlled if self-restraining) and produces different effects:
> * Can handcuff one harm-hand with one-regular hand and upon dropping harmpack having one handcuffed hand and the other free but unusable ❌
> * Cannot use harm-hands if harmpack put on after regular hands restrained ❌ but if another set is put on and removed it works
> * Cannot use any hands if their regular hands get restrained (gets cuffed icon), but can use any free hands if at least one regular hand left unrestrained ❌
> 
> Seems like this in general needs a more thorough fix where restrainment status is divorced from hands usability and both dynamically depend on number of restrained/free hands. In short: restrainment should be per hand, and statuses derived from restrainment (icon, inability to switch hands) should depend on hands' statuses instead of being their own value.

Missing cuffed icon when single-cuffed with harmpack, - this one now should work after new fixes

Which hands end up restrained if more than two are available is random (controlled if self-restraining) and produces different effects - this one should be fixed now, it used to depend on active arm, so if arm 4 was active during cuffing it would take arm 4 and 1

Cannot use harm-hands if harmpack put on after regular hands restrained - yeah you can open issue for that, it's better to not grow this PR any more


Cannot use any hands if their regular hands get restrained (gets cuffed icon), but can use any free hands if at least one regular hand left unrestrained - i am not sure about this one, can't replicate

**Comment by LordCarve** [comment]
Testing:
- ✅ applying handcuffs and removing them via any means works perfectly in my testing (have only tested 2 hands and 2 hands + harmpack cases)
- ✅ unable to replicate the issue of applying handcuffs to odd hands (i.e. one regular, one harmpack). Most likely resolved thanks to `SortedHands`
- ✅ removing harmpack while already handcuffed works OK regardless which set of hands is handcuffed
- ✅ everything alert related-looks good
- ❌ putting harmpack on an already handcuffed person doesn't let them switch or use the newly-acquired hands

**Review by LordCarve — APPROVED** [review]
Code looks OK.
There is one small edge-case issue (see my latest test comment), but author suggests to address it in another PR.

In any case this is an improvement addressing many issues and not introducing new ones. Just because it leaves one edge-case unsolved shouldn't stop it from being merged IMO.
